File: solverAndGenerator/battleships_generator.py
from dataclasses import dataclass
from itertools import chain
from random import sample

from placer import BattleshipsPlacer
from soft_constrains import SoftConstrains
from hard_constrains import HardConstrains
from wave_function_collapse import WaveFunctionCollapse
from states import BattleshipsStates

import numpy as np


# read from json
WATER = (1, {
    (0, 1): {1, 2, 3},
    (0, -1): {1, 2, 3},
    (-1, 0): {1, 2, 3},
    (1, 0): {1, 2, 3},
    (-1, 1): {1, 2, 3},
    (1, 1): {1, 2, 3},
    (-1, -1): {1, 2, 3},
    (1, -1): {1, 2, 3}
})

HOR_SHIP = (2, {
    (0, 1): {1},
    (0, -1): {1},
    (-1, 0): {1, 2},
    (1, 0): {1, 2},
    (-1, 1): {1},
    (1, 1): {1},
    (-1, -1): {1},
    (1, -1): {1}
})

VER_SHIP = (3, {
    (0, 1): {1, 3},
    (0, -1): {1, 3},
    (-1, 0): {1},
    (1, 0): {1},
    (-1, 1): {1},
    (1, 1): {1},
    (-1, -1): {1},
    (1, -1): {1}
})


@dataclass(frozen=True)
class BattleshipsGenerator:
    width: int
    height: int
    ship_lengths: map
    resolution: int = 2

    CONSTRAINS = [WATER, HOR_SHIP, VER_SHIP]
    STATES = BattleshipsStates()

    def __unify(self, board: np.ndarray) -> np.ndarray:
        for y, row in enumerate(board):
            for x, cell in enumerate(row):
                if cell == 3:
                    board[y][x] = 2
        return board

    def __on_board(self, position: tuple) -> bool:
        x, y = position
        return self.width > x >= 0 and self.height > y >= 0

    def __ship_neighbourhood(self, position: tuple) -> list:
        x, y = position
        return [cell for cell in [
            (x, y - 1),
            (x, y + 1),
            (x + 1, y),
            (x - 1, y),
        ] if self.__on_board(cell)]

    def __directed_vector(self, neighbour: tuple, position: tuple) -> int:
        x_diff = position[0] - neighbour[0]
        y_diff = position[1] - neighbour[1]
        if x_diff != 0:
            return 3 if x_diff > 0 else 5
        return 4 if y_diff > 0 else 6

    def __vectorize(self, board: np.ndarray, position: tuple) -> int:
        neighborhood = [(x, y) for x, y in
            self.__ship_neighbourhood(position) if board[y][x] > 1]
        if not neighborhood:
            return 7
        if len(neighborhood) == 1:
            return self.__directed_vector(neighborhood[0], position)
        return 8

    def __strip(self, board: np.ndarray) -> list:
        indexes = list(chain.from_iterable([[(x, y) for x, _ in
            enumerate(row) if board[y][x] > 1
            ] for y, row in enumerate(board)]))
        stripped_board = []
        for x, y in sample(indexes, self.resolution):
            if board[y][x] == 1:
                stripped_board.append((x, y, 1))
            else:
                stripped_board.append((x, y, self.__vectorize(
                    board, (x, y))))
        return stripped_board

    def __calculate_row_lim(self, row: np.ndarray) -> list:
        return len([val for val in row if val == 2])

    def __calculate_limits(self, board: np.ndarray) -> list:
        row_limits = [self.__calculate_row_lim(row)
            for row in board]
        col_limits = [self.__calculate_row_lim(board[:,col])
            for col in range(self.width)]
        return row_limits, col_limits

    def __call__(self, placer: BattleshipsPlacer) -> tuple:
        constants = [None, None, self.ship_lengths]
        soft = SoftConstrains(constants, placer)
        hard = HardConstrains(constants, placer)
        solver = WaveFunctionCollapse(self.CONSTRAINS,
            hard.ship_length_constrain, soft.ship_length_constrain,
            placer, self.STATES)
        board = self.__unify(solver(self.width, self.height, []))
        return self.__strip(board), *self.__calculate_limits(board)

File: src/backend/src/facades/BattleshipsGenerator.java
package backend.src.facades;

import backend.src.boards.Board;
import backend.src.boards.GeneratorBoard;
import backend.src.constrains.Constrains;
import backend.src.constrains.HardConstrains;
import backend.src.constrains.SoftConstrains;
import backend.src.solvers.NoSolutionException;
import backend.src.solvers.WaveFunctionCollapse;
import backend.src.states.BattleshipsStates;
import backend.src.states.States;
import backend.src.utility.Coord;
import backend.src.utility.InitValue;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class BattleshipsGenerator {

    private final Map<Integer, Map<Coord, Set<Integer>>>        constrains;
    private final TreeMap<Integer, Integer>                     shipLengths;
    private final Coord                                         dimensions;
    private final int                                           resolution;

    private final States                                        states = new BattleshipsStates();

    public record Solution(List<InitValue> initValues, List<Integer> rowLimits, List<Integer> columnLimits) { }

    private static class SoftGeneratorConstrains implements Constrains {
        private final SoftConstrains                        constrains;

        public SoftGeneratorConstrains(TreeMap<Integer, Integer> shipLengths) {
            constrains = new SoftConstrains(null, null, shipLengths);
        }

        public boolean check(Board<Set<Integer>> board) {
            return constrains.shipLengthConstrain(board);
        }
    }

    private static class HardGeneratorConstrains implements Constrains {
        private final HardConstrains                        constrains;

        public HardGeneratorConstrains(TreeMap<Integer, Integer> shipLengths) {
            constrains = new HardConstrains(null, null, shipLengths);
        }

        public boolean check(Board<Set<Integer>> board) {
            return constrains.shipLengthConstrain(board);
        }
    }

    public BattleshipsGenerator(Map<Integer, Map<Coord, Set<Integer>>> constrains, TreeMap<Integer, Integer> shipLengths, Coord dimensions, int resolution) {
        this.constrains = constrains;
        this.shipLengths = shipLengths;
        this.dimensions = dimensions;
        this.resolution = resolution;
    }

    private List<Integer> calculateRowLimits(Board<Integer> board) {
        List<Integer> rowLimits = new ArrayList<>();
        for (var row : board) {
            int limit = 0;
            for (Integer value : row) {
                limit += value == 2 ? 1 : 0;
            }
            rowLimits.add(limit);
        }
        return rowLimits;
    }

    private List<Integer> calculateColumnLimits(Board<Integer> board) {
        List<Integer> columnLimits = new ArrayList<>();
        for (var row : board.transpose()) {
            int limit = 0;
            for (Integer value : row) {
                limit += value == 2 ? 1 : 0;
            }
            columnLimits.add(limit);
        }
        return columnLimits;
    }

    private List<Coord> getNeighborhood(Board<Integer> board, Coord position) {
        int x = position.getX(), y = position.getY();
        List<Coord> neighborhood = new ArrayList<>(Arrays.asList(
                new Coord(x, y - 1),
                new Coord(x, y + 1),
                new Coord(x + 1, y),
                new Coord(x - 1, y)
        ));
        neighborhood.removeIf(c-> !board.onBoard(c));
        neighborhood.removeIf(c-> board.accessCell(c) == 1);
        return neighborhood;
    }

    private int directedVector(Coord neighbor, Coord position) {
        int xDiff = position.getX() - neighbor.getX();
        int yDiff = position.getY() - neighbor.getY();
        if (xDiff != 0) {
            return xDiff > 0 ? 3 : 5;
        }
        return yDiff > 0 ? 4 : 6;
    }

    private int vectorize(Board<Integer> board, Coord position) {
        List<Coord> neighborhood = getNeighborhood(board, position);
        if (neighborhood.isEmpty()) {
            return 7;
        }
        return neighborhood.size() == 1 ? directedVector(neighborhood.get(0), position) : 8;
    }

    private List<Coord> generateIndexes(Board<Integer> board) {
        List<Coord> coords = new ArrayList<>();
        for (int x = 0; x < dimensions.getX(); ++x) {
            for (int y = 0;y < dimensions.getY(); ++y) {
                Coord coord = new Coord(x, y);
                if (board.accessCell(coord) != 1) {
                    coords.add(coord);
                }
            }
        }
        return coords;
    }

    private List<InitValue> strip(Board<Integer> board) {
        List<Coord> indexes = generateIndexes(board);
        Collections.shuffle(indexes);
        List<InitValue> initValues = new ArrayList<>();
        for (int i = 0; i < resolution; ++i) {
            initValues.add(new InitValue(indexes.get(i),
                    board.accessCell(indexes.get(i)) == 1 ? 1 : vectorize(board, indexes.get(i))));
        }
        return initValues;
    }

    public Solution generate() throws NoSolutionException {
        SoftGeneratorConstrains soft = new SoftGeneratorConstrains(shipLengths);
        HardGeneratorConstrains hard = new HardGeneratorConstrains(shipLengths);
        WaveFunctionCollapse solver = new WaveFunctionCollapse(soft, hard, constrains, states);
        var result = solver.solve(new GeneratorBoard(dimensions), new ArrayList<>());
        return new Solution(
                strip(result),
                calculateRowLimits(result),
                calculateColumnLimits(result)
        );
    }

}
